### ISSUE TITLE
Catchup with the latest Canon tags from exiftool

### ISSUE THREAD

**Issue by sridharb1** [Exiv2/exiv2#1004]
I am attaching a patch against canonmn_int.cpp that brings exiv2 to almost the latest exiftool Canon.pm (possibly to a couple of weeks ago). Apart from the name differences, there are some space differences in the lens section because the alignment of the numbers 1-999 have only 5 characters before the "," and numbers greater than them have 6 characters before the ",". I standardized on 6 characters before the ",".

With this, you get only 6 failures in the test suite whose output can easily be updated, as opposed 96 differences between exiv2 and exiftool. The advantage with using the patched version is that it is easier to catch up to exiftool afterwards because the comparison would result only in additions (theoretically).

Please note that there are 6 differences that crop up in the test tool whose output need updates to these changed lens names.
[canonmn_int.cpp.patch.zip](https://github.com/Exiv2/exiv2/files/3635564/canonmn_int.cpp.patch.zip)

**Comment by clanmills** [comment]
@sridharb1  Thank for working on this an providing a patch.  The test exceptions are:
```
Files /Users/rmills/gnu/github/exiv2/0.27-maintenance/test/tmp/exiv2-test.out and /Users/rmills/gnu/github/exiv2/0.27-maintenance/test/data/exiv2-test.out differ
1893c1893
< 20001004_015404.jpg   Exif.CanonCs.LensType                        Short       1  Canon EF 28-70mm f/2.8L USM or Other Lens
---
> 20001004_015404.jpg   Exif.CanonCs.LensType                        Short       1  Canon EF 28-70mm f/2.8L
3505c3505
...
```

And:
```
FAIL: test_run (bugfixes.github.test_issue_20.TamronSupport)
----------------------------------------------------------------------
- Exif.CanonCs.LensType                        Short       1  254
+ Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD Macro 1:1

FAIL: test_run (bugfixes.github.test_issue_45.Sigma24_105mmRecognization)
----------------------------------------------------------------------
- Exif.CanonCs.LensType                        Short       1  143
+ Exif.CanonCs.LensType                        Short       1  Sigma 24-105mm F4 DG OS HSM [Art 013]

- Exif.CanonCs.LensType                        Short       1  173
+ Exif.CanonCs.LensType                        Short       1  Sigma 150-500mm f/5-6.3 APO DG OS HSM + 2x

FAIL: test_run (bugfixes.redmine.test_issue_1252.CanonLenses)
----------------------------------------------------------------------
- Exif.CanonCs.LensType                        Short       1  Sigma APO 120-300mm f/2.8 EX DG OS HSM
+ Exif.CanonCs.LensType                        Short       1  Sigma 120-300mm f/2.8 EX APO DG OS HSM
```

Later today, I'll fix up the test harness (to report success) and submit a PR for team review.

**Comment by clanmills** [comment]
I'm working on the patch.  This isn't what I expected.  I thought we would now recognise:
```
254 => Tamron ....
```
However it's the other way about:
```
.../build $ bin/exiv2 --grep lenstype/i ../test/data/exiv2-g20.exv   # build with your patch
Exif.CanonCs.LensType                        Short       1  254
.../build $ exiv2 --grep lenstype/i ../test/data/exiv2-g20.exv # current 0.27-maintenance patch
Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD Macro 1:1
.../build $ 
```

For what reason should we stop recognising Exif.CanonCs.LensType 254 as Tamron SP 90mm f/2.8 Di VC USD Macro 1:1 ?

**Comment by sridharb1** [comment]
There are two 254s as you can see in the old file. The 254.1 (Sigma) is not there in the latest exiftool. I leave it up to you to preserve it.
Regards,Sridhar

    On Friday, 20 September, 2019, 7:15:59 pm IST, Robin Mills <notifications@github.com> wrote:  
 
 
I'm working on the patch. This isn't what I expected. I thought we would now recognise:
254 => Tamron ....

However it's the other way about:
.../build $ bin/exiv2 --grep lenstype/i ../test/data/exiv2-g20.exv   # build with your patch
Exif.CanonCs.LensType                        Short       1  254
.../build $ exiv2 --grep lenstype/i ../test/data/exiv2-g20.exv # current 0.27-maintenance patch
Exif.CanonCs.LensType                        Short       1  Tamron SP 90mm f/2.8 Di VC USD Macro 1:1
.../build $ 

For what reason should we stop recognising Exif.CanonCs.LensType 254 as Tamron SP 90mm f/2.8 Di VC USD Macro 1:1 ?

**Comment by clanmills** [comment]
Thanks for the explanation.  I've "tweaked" several other "duplicated IDs".  Here's the difference between your patch and PR #1005.
```
710 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ diff ../src/canonmn_int.cpp ~/temp/canonmn_int.cpp 
898a899
>         {  136, "Tamron SP 15-30mm f/2.8 Di VC USD A012"                    }, // 1
923a925
>         {  143, "Sigma 24-105mm F4 DG OS HSM [Art 013]"                     }, // 2
999,1001c1001,1004
<         {  173, "Sigma APO Macro 150mm f/2.8 EX DG HSM"                     }, // 2
<         {  173, "Sigma 10mm f/2.8 EX DC Fisheye"                            }, // 3
<         {  173, "Sigma 15mm f/2.8 EX DG Diagonal Fisheye"                   }, // 4
---
>         {  173, "Sigma 150-500mm f/5-6.3 APO DG OS HSM + 2x"                }, // 2
>         {  173, "Sigma APO Macro 150mm f/2.8 EX DG HSM"                     }, // 3
>         {  173, "Sigma 10mm f/2.8 EX DC Fisheye"                            }, // 4
>         {  173, "Sigma 15mm f/2.8 EX DG Diagonal Fisheye"                   }, // 5
1125a1129
>         {  254, "Tamron SP 90mm f/2.8 Di VC USD Macro 1:1"                  }, // 1
711 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

**Comment by sridharb1** [comment]
Ok, thanks.

    On Friday, 20 September, 2019, 9:29:05 pm IST, Robin Mills <notifications@github.com> wrote:  
 
 
Thanks for the explanation. I've "tweaked" several other "duplicated IDs". Here's the difference between your patch and PR #1005.
710 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ diff ../src/canonmn_int.cpp ~/temp/canonmn_int.cpp 
898a899
>         {  136, "Tamron SP 15-30mm f/2.8 Di VC USD A012"                    }, // 1
923a925
>         {  143, "Sigma 24-105mm F4 DG OS HSM [Art 013]"                     }, // 2
999,1001c1001,1004
<         {  173, "Sigma APO Macro 150mm f/2.8 EX DG HSM"                     }, // 2
<         {  173, "Sigma 10mm f/2.8 EX DC Fisheye"                            }, // 3
<         {  173, "Sigma 15mm f/2.8 EX DG Diagonal Fisheye"                   }, // 4
---
>         {  173, "Sigma 150-500mm f/5-6.3 APO DG OS HSM + 2x"                }, // 2
>         {  173, "Sigma APO Macro 150mm f/2.8 EX DG HSM"                     }, // 3
>         {  173, "Sigma 10mm f/2.8 EX DC Fisheye"                            }, // 4
>         {  173, "Sigma 15mm f/2.8 EX DG Diagonal Fisheye"                   }, // 5
1125a1129
>         {  254, "Tamron SP 90mm f/2.8 Di VC USD Macro 1:1"                  }, // 1
711 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $

**Comment by clanmills** [comment]
@sridharb1 Can you submit a PR for this change?  I will help you with the git monster.

I've dealt with every known issue for Exiv2 v0.27.3.  I thought it would take weeks, but I'm almost at RC1 after one week.  My original idea was to publish RC1 without ISOBMFF support on April 30.  I've made such good progress, I will be able to include ISOBMFF support in RC1.  It's better to be feature complete at RC1.

**Comment by sridharb1** [comment]
Great progress Robin. Let me see how I can help in this regard.

**Comment by clanmills** [comment]
@sridharb1 Exiv2 v0.27.3 is fast approaching RC1.  It's now 88% complete.  If you wish to these Canon changes included, we should cooperate on this in the next week.

My proposals about ISOBMFF have been met with hostility and I will do no further work on ISOBMFF.  https://github.com/Exiv2/exiv2/issues/1066#issuecomment-610227240

My other tasks for Exiv2 v0.27.3 are quite small and I hope to reach RC1 in the next week.

**Comment by sridharb1** [comment]
I have caught up to the latest released exiftool, v11.93 with these updates.

Specific changes from last time:

1. I have not blindly made the lens name same as that in exiftool. For e.g., if a lens is called "Canon EF 28mm f/2.8 or Sigma Lens" in exiftool and there is a distinguishing lens with the same id (in this case, "Sigma 24mm f/2.8 Super Wide II"), I remove the "or Sigma lens" from the name. Explanation: I happen to own the "Canon EF 24-70mm f/2.8L II USM". It is a great lens that I paid a lot of money for. When I use exiv2 to tag the photos that I take with this lens, I would like to show that it was taken with "Canon EF 24-70mm f/2.8L II USM" and not "Canon EF 24-70mm f/2.8L II USM or Sigma lens". I would assume that this would be the same with other people too. What Sigma and the other manufacturers who make third party lenses do is they try to replicate the original in almost all respects including the code that it emits. That's why you have these "duplicates" sometimes. Sigma has stopped doing that with its newer lenses, which is great. But one shouldn't have to "dilute" the original lens just because there is another one professing to be the same as long as it is identifiable through a separate code.
2. I have removed some of the codes attached to Tamron lenses, for e.g. in exiftool, "Tamron SP 70-200mm f/2.8 Di VC USD G2 (A025)" is the exiftool name, but the trade name of the lens is "Tamron SP 70-200mm f/2.8 Di VC USD G2" and Tamron's internal model number is A025. Again, people would want their lens identified using its trade name and not have these model numbers appended to them.

I used an excel spreadsheet to create these. I am attaching that in case others want to extend the same to other manufacturers or catch up to a newer version. Apart from excel, it uses emacs to do some regex replacements, but the same can be done through another tool as well.

There may be some squeaks from tests, but I have not looked at them yet.

[canon models and lenses.xlsx](https://github.com/Exiv2/exiv2/files/4445808/canon.models.and.lenses.xlsx)
[canonmn_int.cpp.2.patch.txt](https://github.com/Exiv2/exiv2/files/4445809/canonmn_int.cpp.2.patch.txt)

**Comment by clanmills** [comment]
Thanks for this.  Your patch is good and (almost) runs passes the test suite.  I'll fix the tests and put your change into sridharb1_contrib_msvc4_0.27 this evening.

I work on the Mac.  I can install Apple's Numbers.app which I believe will open the spreadsheet.  If not, I think I've got a copy of Excel somewhere.

**Comment by clanmills** [comment]
I'll pushed your code change into branch `sridharb1_contrib_msvc4_0.27`.  You can sync with it:
```
$ git checkout sridharb1_contrib_msvc4_0.27
$ git pull --rebase
```
I've updated test/data/exiv2-test.out, which correctly detect this change:
```
< > 20001004_015404.exv   Exif.CanonCs.LensType                        Short       1  Canon EF 28-70mm f/2.8L USM
---
> > 20001004_015404.exv   Exif.CanonCs.LensType                        Short       1  Canon EF 28-70mm f/2.8L USM or Other Lens
```

I'd like to ask which lens 173 is no longer identified.  It used to be `Sigma 150-500mm f/5-6.3 APO DG OS`.  Similarily 143 was `Sigma 24-105mm F4 DG OS HSM [Art 013]` and 254 was `Tamron SP 90mm f/2.8 Di VC USD Macro 1:1`

```
569 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ $EXIV2_BINDIR/exiv2 -pa -g Lens data/exiv2-bug1122.exv 
Exif.CanonCs.LensType                        Short       1  173
Exif.CanonCs.Lens                            Short       3  300.0 - 1000.0 mm
Exif.Canon.LensModel                         Ascii      74  300-1000mm
Exif.Photo.LensSpecification                 Rational    4  300/1 1000/1 0/1 0/1
Exif.Photo.LensModel                         Ascii      11  300-1000mm
Exif.Photo.LensSerialNumber                  Ascii      11  0000000000
570 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test $ exiv2 -pa -g Lens data/exiv2-bug1122.exv 
Exif.CanonCs.LensType                        Short       1  Sigma 150-500mm f/5-6.3 APO DG OS HSM + 2x
Exif.CanonCs.Lens                            Short       3  300.0 - 1000.0 mm
Exif.Canon.LensModel                         Ascii      74  300-1000mm
Exif.Photo.LensSpecification                 Rational    4  300/1 1000/1 0/1 0/1
Exif.Photo.LensModel                         Ascii      11  300-1000mm
Exif.Photo.LensSerialNumber                  Ascii      11  0000000000
```
I haven't looked at your spreadsheet (it's been another long day).  When we're done with this, I'll store it in the "team" SVN repos.  We use the "team" repos to keep documents, the code to generate the web-site and so on.  There's nothing secret in "team", however I had complaints about our code repos being too large and there is no reason for it to be part of the code repos.

**Comment by sridharb1** [comment]
Sigma 150-500mm f/5-6.3 APO DG OS is now in 174. Sigma 24-105mm F4 DG OS HSM is in 255 and 
Tamron SP 90mm f/2.8 Di VC USD Macro 1:1 is in 26.2. May not be the exact same name, but most likely the "correct" name.

Maybe exiftool changed the codes for some reason.

**Comment by clanmills** [comment]
@sridharb1 Good News: I've installed Numbers.app and it opens your spreadsheet.  Boy, that regex looks the business!

We have to reconcile this.  We have three tests which are broken by your changes.

| Issue | Test images | Bug Report |
|-- |-- |-- |
| g20 |         exiv2-g20.exv<br>CanonEF100mmF2.8LMacroISUSM.exv<br>TamronSP15-30mmF2.8DiVCUSDA012.exv<br>TamronSP90mmF2.8DiVCUSDMacroF004.exv<br>TamronSP90mmF2.8DiVCUSDMacroF017.exv | https://github.com/Exiv2/exiv2/issues/20 |
| g45 | exiv2-g45.exv | https://github.com/Exiv2/exiv2/issues/45 |
| r1122 | exiv2-bug1122.exv | https://dev.exiv2.org/issues/1122 |

Looking at at g45 
```
.../test/data $ exiv2 -pa --grep lens/i *g45*
Exif.CanonCs.LensType                        Short       1  Sigma 24-105mm F4 DG OS HSM [Art 013]
Exif.CanonCs.Lens                            Short       3  24.0 - 105.0 mm
Exif.CanonCf.LensAFStopButton                Short       1  0
Exif.Canon.LensModel                         Ascii      74  24-105mm F4 DG OS HSM | Art 013
.../test/data $ exiv2 -pv --grep lens/i *g45*
0x0016 CanonCs      LensType                    Short       1  143
0x0017 CanonCs      Lens                        Short       3  105 24 1
0x0009 CanonCf      LensAFStopButton            Short       1  0
0x0095 Canon        LensModel                   Ascii      74  24-105mm F4 DG OS HSM | Art 013
.../test/data $ exiftool *g45* | grep -i lens
Lens Type                       : Canon EF 500mm f/4L IS USM or Sigma Lens
Lens Model                      : 24-105mm F4 DG OS HSM | Art 013
Lens                            : 24.0 - 105.0 mm
Lens ID                         : Sigma 17-70mm f/2.8-4 DC Macro OS HSM + 1.4x
Lens                            : 24.0 - 105.0 mm (35 mm equivalent: 23.7 - 103.7 mm)
$ 
```
So, we currently report for 143 "Sigma 24-105mm F4 DG OS HSM [Art 013]".  With your change we get:
```
.../test/data $ ../../build/bin/exiv2 -pa --grep lens/i *g45*
Exif.CanonCs.LensType                        Short       1  143
Exif.CanonCs.Lens                            Short       3  24.0 - 105.0 mm
Exif.CanonCf.LensAFStopButton                Short       1  0
Exif.Canon.LensModel                         Ascii      74  24-105mm F4 DG OS HSM | Art 013
```
I'm lost and confused.  I haven't investigated g20 and r1122 because I want to work on Unix support today. 

Can you investigate these three puzzles, please.

**Comment by sridharb1** [comment]
I will look into it. It looks like #1122 is a special case which would necessitate the reinsertion of the old line dealing with Sigma.

We may have some issue with the Lens Type line vs Lens Model vs Lens ID. Let me investigate further.

**Comment by clanmills** [comment]
@sridharb1 Lens identification is a mess.  I wish Andreas had never got into this.  It's the source of about 50% of issues.    Here's the raw data in 1122:
```
.../test/data $ exiv2 -pv --grep lens/i exiv2-bug1122.exv 
0x0016 CanonCs      LensType                    Short       1  173
0x0017 CanonCs      Lens                        Short       3  1000 300 1
0x0095 Canon        LensModel                   Ascii      74  300-1000mm
0xa432 Photo        LensSpecification           Rational    4  300/1 1000/1 0/1 0/1
0xa434 Photo        LensModel                   Ascii      11  300-1000mm
0xa435 Photo        LensSerialNumber            Ascii      11  0000000000
654 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/test/data $ 
```
And when it's interpreted (`-pt, or -pa`):
```
.../test/data $ exiv2 -pa --grep lens/i exiv2-bug1122.exv 
Exif.CanonCs.LensType                        Short       1  Sigma 150-500mm f/5-6.3 APO DG OS HSM + 2x
Exif.CanonCs.Lens                            Short       3  300.0 - 1000.0 mm
Exif.Canon.LensModel                         Ascii      74  300-1000mm
Exif.Photo.LensSpecification                 Rational    4  300/1 1000/1 0/1 0/1
Exif.Photo.LensModel                         Ascii      11  300-1000mm
Exif.Photo.LensSerialNumber                  Ascii      11  0000000000
.../test/data $ 
```
**Exif.Canon.LensModel, Exif.Photo.LensModel and Exif.Photo.LensSerialNumber** are `Ascii`.  _**That's good!  Canon wrote that ascii data.  Don't touch it**_

**Exif.Photos.LensSpecification** are for `Rational` numbers: 300, 1000,0,0 and the interpretation "300-1000mm" seems OK.  **Exif.CanonCs.Lens** is similar and comes from the MakerNote.

**Exif.CanonCs.LensType** is a `Short` value = 173.  That's what we're discussing.  Comes from the MakerNote.

These "magic numbers" are in different places in the different manufacturers makernotes.  This seriously sucks.

The purpose of the file: ~/.exiv2 is to give the user the ability to change the "interpretation" for an ID and I believe it works well.

I appreciate your efforts here to sync our recognition with ExifTool.  Thanks for working on this.

**Comment by sridharb1** [comment]
Ok, I have made some small changes. Now, all testcases pass. Here's the patch that I had to do get the test scripts to work and some datafile modifications. The exiv2-test.out changes, I believe, you have already done.

Instead of the hardcoded .exe that I added to functions.source, you can use the variable $EXIV2_EXT that you suggest for the python tests.

Running the python tests from cygwin (as part of make test) is probably not advised for Windows. It is better off run through a cmd prompt. 

This is how I ran it.

set EXIV2_EXT=.exe
set EXIV2_BINDIR=E:\Projects\exiv2\contrib\vs2019\solution\src\x64\Debug
"C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\python.exe" runner.py --verbose 1>tests.out 2>tests.err

The unix tests all succeed.

[unix-tests.txt](https://github.com/Exiv2/exiv2/files/4451059/unix-tests.txt)

The output of the python tests are attached. One of them fails, but maybe you can see why.
[canonmn_int.cpp.3.patch.txt](https://github.com/Exiv2/exiv2/files/4451029/canonmn_int.cpp.3.patch.txt)

[tests.err.txt](https://github.com/Exiv2/exiv2/files/4451041/tests.err.txt)

**Comment by clanmills** [comment]
You're a clever guy to figure out how to run the test stuff on your own.  It is documented in README.md.  Good idea to use EXIV2_EXT - Dan added that for the python tests.  It's never been in functions.source.  Another part of Exiv2 v0.28 was to use CTest to make our testing platform independent.  It's currently a jumble of stuff going back over 15 years.  However, it's stable and works.

Your patch smells right.  However git on the Mac barfs and won't apply it.

Can you zip and send me the files you've changed.  And yes, I will run dos2unix on them.
```
 src/canonmn_int.cpp                    | 174 ++++++++++++++++++---------------
 test/data/exiv2-test.out               |  10 +-
 test/functions.source                  |   2 +-
 test/path-test.sh                      |   6 +-
 test/xmpparser-test.sh                 |   1 +
 tests/bugfixes/github/test_issue_20.py |   4 +-
 tests/bugfixes/github/test_issue_45.py |   2 +-
```

**Comment by clanmills** [comment]
You can safely ignore the output in tests.err.txt on #987

That test has been removed from the test suite.  It's purpose is to see if the output of samples/tagname has changed.  And of course it has for your situation as you have modified the tag tables in tags_int.cpp.  Ignore this.  Remove the file `tests/bugfixes/github/test_issue_987.py` and it won't get run.

**Comment by sridharb1** [comment]
> Your patch smells right. However git on the Mac barfs and won't apply it.

Don't use git to patch if you can. The problem is that the diff for canonmn_int.cpp is against the original and thus, the huge number of changes (174), whereas I pasted only the real differences from the file that I had sent yesterday. I would patch each of these by hand. Or at the very least, remove the canonmn part from the patch, patch that by hand and patch the remainder by git.

**Comment by clanmills** [comment]
OK.  Sounds like pain - however it won't kill me.  While I fix canonmn_int.cpp by hand can you send a patch for the other files.  I think you can get that with:

```
$ git diff test/data/exiv2-test.out  test/functions.source test/path-test.sh test/xmpparser-test.sh tests/bugfixes/github/test_issue_20.py tests/bugfixes/github/test_issue_45.py > foo.patch
```
We'll be done in 10 minutes.

**Comment by sridharb1** [comment]
Try this with git. Should patch all files.
[canonmn_int.cpp.4.patch.txt](https://github.com/Exiv2/exiv2/files/4451687/canonmn_int.cpp.4.patch.txt)

**Comment by clanmills** [comment]
I think I've almost got it done with the earlier stuff.  It's getting late for you.  We'll talk tomorrow.

**Comment by clanmills** [comment]
Right.  I've got that committed.  I made some little changes: 1496880

I'd like you to use git.  I know there's learning involved, however I'll help.  I'd prefer to give you help with git that be manually patching files.  You might even like it.  And it's a chance for to test my git knowledge to somebody else.   Kind of proof that I understand git.

Comments.

1) I see you changed $exe to .exe in `functions.source`.  The wrecks the test suite on the Mac.  I've added code about `if [ ! -z $EXIV2_EXT ] ; then ext=$EXIV2_EXT ; fi`

2) I've changed the index markers on entry 254 in canonmn_int.cpp

3) I've commented off `unixtodos` because that again wrecks `xmpparser.sh`.  I'm not too sure why you need this.  As things stand, the diff report is produced by calling `diff` with `d--strip-trailing-cr`.

4) You've change `path-test` to `pathtest`.  Is there a reason for this?  The code is src/path-test.cpp

We're going well.  When I persuade you to use git, we'll move even quicker.

I didn't make any progress with UNIX today - loads of other interruptions.  If you agree to work with git tomorrow, I'll spend time getting all the dependences for `contrib/vs2019` into place.

**Comment by sridharb1** [comment]
The exiv2_ext changes are good.

The canonmn_int.cpp changes seem fine.

At least for me, I need the unix2dos in xmpparser.sh. Otherwise, I get spurious output when really only trivial stuff has changed.

path-test, the project is called pathtest now and its output, by default, is pathtest.exe. If it should have been path-test.exe, it would be nice for the project to be path-test.vcxproj. Don't know where this changed.

I am working with git. So, let's see how to make things as easy as possible for you.

**Comment by clanmills** [comment]
@sridharb1 I know what to do about xmpparser-test.sh/unix2dos.   It's calling `diff` directly without using `----strip-trailing-cr`.  The utility `diff` should be called as `diff $diffargs` in the test suite.

A bit of history.  Those bash scripts were all totally different and written by different people.  About 2012, I invented `functions.source` to provide a bash library which are read by `source functions.source` in test/*.sh.  Then I tidies up the code.  By the time I arrived (in alphabetical order) at xmpparser-test.sh, I was bored and I missed that.  

Git:
_1 **Clone our branch and do some work**_

```
$ git clone https://github.com/exiv2/exiv2 --branch sridharb1_contrib_msvc4_0.27 work
$ cd work
------ do whatever ----
$ git status
---- find out what you've done ---
$ git add file1 file.....
$ git commit --message "....."
```

_2 **Push changes from your machine to GitHub**_
```
$ git branch --set-upstream-to origin/sridharb1_contrib_msvc4_0.27
$ git push 
```
Check this page:  https://github.com/Exiv2/exiv2/pull/1136 and you'll see your push.  The CI tools will run (takes a couple of hours for the builders to be assigned and run).

_3 **When you want to pull changes that I've made onto your machine**_
```
$ cd work
$ git pull --rebase
```

_4 **To make changes and push them to GitHub**_
```
$ cd work
$ git pull --rebase
--------- edit -----
$ git commit --message "............"
$ git push
```

No more patches.  Git does it for you.  I use SmartGit so that I don't need to know the commands.
